File: problem/Palindrome_Number/isPalindrome.py
class Solution(object):
    def isPalindrome(self, x):
        """
        Mathematical approach - O(log n) time, O(1) space
        """
        if x < 0 or (x % 10 == 0 and x != 0):
            return False
        
        reversed_half = 0
        while x > reversed_half:
            reversed_half = reversed_half * 10 + x % 10
            x //= 10

        return x == reversed_half or x == reversed_half // 10

    def isPalindrome_string(self, x):
        """
        String conversion approach - O(log n) time, O(log n) space
        """
        if x < 0:
            return False
        
        s = str(x)
        left, right = 0, len(s) - 1
        
        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        
        return True

File: problem/Palindrome_Number/isPalindrome.c
bool isPalindrome(int x)
{

    if (x < 0 || (x % 10 == 0 && x != 0))
    {
        return false;
    }

    int reversedHalf = 0;
    while (x > reversedHalf)
    {
        reversedHalf = reversedHalf * 10 + x % 10;
        x /= 10;
    }

    return x == reversedHalf || x == reversedHalf / 10;
}

bool isPalindrome_string(int x)
{

    if (x < 0)
        return false;

    char s[12];
    sprintf(s, "%d", x);

    int len = strlen(s);
    int left = 0, right = len - 1;

    while (left < right)
    {
        if (s[left] != s[right])
        {
            return false;
        }
        left++;
        right--;
    }

    return true;
}
